### ISSUE TITLE
Accelerate `Fl_Text_Display` and `Fl_Text_Buffer`

### ISSUE THREAD

**Issue by MatthiasWM** [fltk/fltk#596]
This is a personal reminder to check if appending many short strings to `Fl_Text_Buffer` can be accelerated.

Also, when utf'ifying `Fl_Text_Display` I seem to remember that there was a way to reduce the time to calculate `Fl_Text_Display::warp()` substantially, but I never implemented that. So, now is the time (hint: binary search).

**Comment by erco77** [comment]
See also related issue #300.

**Comment by MatthiasWM** [comment]
Ok, so I dug through Fl_Text_Display, and I am sorry to say that I can't accelerate that without a wide ranging rewrite, and I wouldn't;t even know if the result is better under all circumstances. My acceleration idea works, but wasn't the time waster I suspected after all. Disabling `wrap()` is the one true accelerator, and if Fl_Simple_Terminal is limited to appending text, inserting a `\n` every 80 or so characters would give a typical terminal effect. But it is somewhat of a hack.

**Comment by erco77** [comment]
@MatthiasWM Thanks for taking a look. Yeah, no one seems to want to touch Fl_Text_xxx stuff, lol.
I may have to do something like disable wrap() for the test programs, as some really are unusable due to the slowness of FST. test/tree and opening a large menu and then doing a large selection shows the problem, such as doing a shift-select over a hundred or so lines repeatedly.

**Comment by erco77** [comment]
OK, the problem with disabling `wrap_mode()` is it causes Fl_Text_Display to force showing a horizontal scrollbar on the screen at all times (whether it needs one or not) once the vertical scrollbar appears.

Which in the case of the all the test apps is kinda a bad problem because the terminal windows at the default app size are necessarily small (just a few lines) the hscrollbar takes up a large percentage of the display and is kinda distracting.

I have an XXX comment in the FST constructor about this:
```
  :
  // XXX: We use WRAP_AT_BOUNDS to prevent the hscrollbar from /always/
  //      being present, an annoying UI bug in Fl_Text_Display.
  wrap_mode(Fl_Text_Display::WRAP_AT_BOUNDS, 0);
  :
```

I think I considered trying to hide the hscrollbar in FTD, but I don't think the widget is designed for that, so had to fallback to enabling wrap to prevent it.

Perhaps the code can be modified so that if the hscrollbar is hidden, it will not try to resize the text area, and can perhaps even save itself some calculation time.

**Comment by MatthiasWM** [comment]
That is a relatively easy fix because it does not display a scrollbar in any of the other modes either, but I am afraid that all the speed advantage will be lost, because calculating the pixel width of every line is exactly the problem that we are seeing in the other modes. Still pondering what we could do to accelerate all this.

**Comment by Albrecht-S** [comment]
> calculating the pixel width of every line is exactly the problem that we are seeing in the other modes. Still pondering what we could do to accelerate all this.

I have a vague idea, off the top of my head. IIRC we'e calculating the line width by calculating every char's width and summing this up. This is very wasteful. IIRC (again) Manolo added caching of single character widths somewhere in the font handling code which helped but is still not satisfactory.

If these two assumptions are true we could instead measure longer strings or at least "words" and then find the wrapping point. If necessary we could use a "binary search" to find the wrapping point. This might be measuring parts of the line more than once, but my assumption is that measuring a string (word) at once is cheaper than measuring all of its characters one by one. Measuring a word would likely be more accurate and since we're only wrapping at word boundaries we wouldn't lose anything.

But if I'm talking nonsense, forget this...

**Comment by Albrecht-S** [comment]
Regarding "binary search": instead of measuring single words we could find the word delimiters (mainly spaces) and measure entire strings, maybe the entire line first. If it fits the width we're ready. If not, use half of the words as a total string and so on. Or use the number of words that **might** fit by interpolating, e.g. if the total line is 150% of the line width we'd use 2/3 of the words and so on. Just thinking...

**Comment by erco77** [comment]
Caching the entire line width per line might be good. Then it's just a short walk of the lines to determine line width.  Whenever a line is modified/edited, that line's "width cache" is updated.

And if the largest width is needed for the horiz scrollbar calc, a running 'maxwidth' could be maintained (since that's all the horiz scroll would care about) and updated whenever a line is modified.

**Comment by erco77** [comment]
Oh, wait, is the slowness we're talking about only when a large number of lines is being `append()`ed, and therefore during the `append()`, the string has to be walked to calc the char's width? I see, if that's the problem, then I see where manolo's fix would help if it's a char width cache for that font. That way if the width of a "w" is calculated once, it wouldn't have to be re-calced. Perhaps such an optimization just for ASCII chars would speed things up, i.e. a small width cache for chars 0x20 thru 0x7f would just be 92 unsigned ints per font.

In my case everything added to the buffer is ASCII in a fixed width font (the test program's text output), so I would think another way to speed up width calc: `if ( font == fixed width && char <= 0x7f )` then the width is the fixed width of the font, and doesn't even have to be calc'ed or cached, helping it zoom through the per-char width calc really fast.

**Comment by erco77** [comment]
Matt -- it might be fun to chat about this, either by voice (facetime) or webchat. If you have time today, lemme know by email.

**Comment by MatthiasWM** [comment]
I don't have time tonight, but we should really try to get some live chatting going. I have a Zoom account, and will try to set something up.

**Comment by erco77** [comment]
Sounds good -- email me with at least 6 hr lead time when you see a free time for this.
I'm looking at the code now, to see if my fixed width fonts + ascii chars optimization can help.

**Comment by erco77** [comment]
OK, one interesting data point: I've come to focus on `Fl_Text_Display::string_width()` which calls `fl_width()` to calculate width in the current font style, and found it's being called way more than it should.

I added a call counter to this function and to `draw()`; using the `test/tree` application and watching the small 8 line FST window: when you run the app it adds a single 71 character string to the terminal, and draw() is called only once. Meanwhile, `string_width()` was called **2,374 times**. 

- The full 71 char string's width is re-calculated repeatedly 31 times
- The _individual characters_ in the string are called all the rest (2,343)

Yeow.
The call frequency seems really excessive just to add a single line and draw it, and 31 repeat calls for the entire string's width seems odd, not to mention all those single char width counts.
That ends up being a lot of calls to `fl_font()` and `fl_width()`.

From there I found:
- Adding 12 lines to the buffer, call count was >80,000.
- Adding another 10 lines: >405,000
- Another 10: >665,000
- Another 10: >980,000
- etc

Considering only 7 lines are visible in the display, most of those calculations must be offscreen at a certain point, recalc'ing the entire buffer over and over..

For laffs, I tried short circuiting that function completely: since `set_width()` seems to be invoked in my case to either to measure single chars or the entire line, and since my lines are all about the same length, I rewrote the function to just do this:
```
double Fl_Text_Display::string_width( const char *string, int length, int style ) const {
  return (length==1) ? 10 : (10*71);   // guess single char width is 10, line width is 10*71
}
```
Surprisingly that crazy hack didn't affect text drawing at all; text appeared properly and scrolled properly. Just the cursor position ended up being a little off from where it should be. As lines went off the screen, the vertical scrollbar did what it should.

That sped things up a little more than 1/2, but even with that crazy short-circuit, it's still quite slow:
- Without the short circuit, adding 500 lines takes 8 secs
- With the short circuit, adding 500 lines takes 3 secs

So there must be really be a lot of excessive stuff going on, wow.

**Comment by erco77** [comment]
FWIW, I'm investigating if I can re-implement `Fl_Simple_Terminal` using its own code to handle the terminal, so that I can decouple from `Fl_Text_Display`.

I've spent about three days working on this, starting from scratch in this [branch](https://github.com/erco77/fltk/tree/Fl_Terminal) in my fltk fork.

I've got what I think is a solid internal framework for efficient management of a utf8 based screen buffer with per-character attributes. My first goal was just to be able to get some lines of text on the screen, and reached that this morning. Added utf8, colors, and simple cursor motion today; kinda happy with the results:
![image](https://user-images.githubusercontent.com/6484779/208813245-f7b82153-50aa-477b-8afd-29e5999867df.png)

So far it's only 350 lines of actual code (according to cloc(1)), so it's a lot smaller than I would have expected. To be seen how efficient it is, but I have high hopes that it can handle bursts of high speed output without slowing things down.

Each character on the display is an instance of a Utf8Char class, which carries the per-character color (a uchar) and attribute (another uchar), and a I think a decent way to manage memory efficiently.

To be done: ESC code processing, screen history, mouse selection, scrollbars. I **don't** plan to implement termio or pty stuff, as I think that can be added by subclassing outside of FLTK. But I have high hopes for creating a small and efficient terminal without subclassing `Fl_Text_Display`.

**Comment by Albrecht-S** [comment]
@erco77 Looks great, I'd like to view the code and test, but ...
> in [this branch](https://github.com/erco77/fltk/tree/Fl_Terminal) of my fltk fork.

... the file `src/Fl_Terminal` is missing. Could you please check it in?

```
$ git diff --stat `git merge-base master HEAD`
 FL/Fl_Terminal.H      | 726 +++++++++++++++++++++++++++++++++++++++++++++++++++++++
 examples/testterm.cxx | 116 +++++++++
 2 files changed, 842 insertions(+)
```

> So far it's only 350 lines of actual code (according to cloc(1)) ...

Hmm, I believe that's not entirely correct. Is this only the (missing) file `src/Fl_Terminal`? What I see now is:
```
$ cloc FL/Fl_Terminal.H
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C/C++ Header                     1             58            171            497
-------------------------------------------------------------------------------
```

But anyway, it's still a great achievement and a lot less than `Fl_Text_Display` and `Fl_Text_Buffer` have together (w/o `Fl_Text_Editor`): header + source = 4034 loc.

**Comment by erco77** [comment]
> ... the file src/Fl_Terminal is missing. Could you please check it in?

Ya, I realized that late last night after I'd headed for bed -- just commited it now, see 83ff0f515.
That should build and run -- that was the state of things when I went to bed.
    
The escape code is untested -- I only had enough energy before sleep to just get the code building again with the new addition. I'll add more today.

>> So far it's only 350 lines of actual code (according to cloc(1)) ...
>
> Hmm, I believe that's not entirely correct.

Heh, at the time I wrote it it was true; _everything_ was in the one .H file, there was no .cxx.
I think the last commit that had everything in the .H was commit 85b01139, which should have built.
    
After that "350" post I kept chugging along for many hours, pulling in the new escape sequence class I'd written for Fl_Simple_Terminal recently. Adding _that_ forced me to split into two files, just to prevent madness.

At some point I'll split everything up nicely, but I'm still in early dev..

**Comment by erco77** [comment]
Oh, and my edit/build cycle has been (assuming Fltk has already been built with configure/make):

    touch testterm.cxx && make testterm && ./testterm
    vi ../*/Fl_Terminal.{cxx,H}

Since everything for this new widget is pulled in with an #include during dev, I don't have to rebuild FLTK, just keep touching/rebuilding the app.

**Comment by erco77** [comment]
I've made a separate issue for this development in my fork, to prevent noise in Matt's issue:
https://github.com/erco77/fltk/issues/1
Anyone interested in tracking/discussing this new widget development can follow up there.

**Comment by MatthiasWM** [comment]
>Meanwhile, `string_width()` was called **2,374 times**.

Yes. If you have wrap mode disabled, it counts the lines by counting '\n' and derives all other information, top displayed line, etc. from that.

If wrap mode is set to the widget border it must:
 - go through the entire document and look at each line
 - it must check the line length in pixels and find out if and where it wraps
 - it must take the remainder of the line and see it that wraps at the right border, and repeat that until the line is done
 - once it know the total number of displayed and possibly wrapped lines, it checks if the vertical scroll bar must be visible (more lines that it can display)
 - if the visibility of the vertical scrollbar changes, the width of the widget changes, so it must go back to the start and measure all lines again
 - now it needs to know the line number of the visible block and where the cursor is, so, um, you guessed it, it measures very line again until it find the top line
 - now it renders all lines in the display by yet again calculating all the wrap positions

Luckily, many of the insert and remove functions are caching those values, so that some operations do not need this amount of work. There is one method though that is not cached and cause a complete recalculation: any for of scrolling! If you grab the scroll bar in editor with a large document loaded (I recommend the Hitchhikers Guid to the Galaxy: it's open source, has a good length for testing, is a fine read, and has not UTF-8 characters), even on my MacBook Pro, it takes almost half a second to start the scrolling.

**Comment by wcout** [comment]
I just tested with "Hitchhikers.." (It has >41780 lines and is about 1.5MB) on Linux/Xft in editor with Wrap mode on.
With FLTK 1.4 current head it behaves as you describe above => pritically unusable.

**But with `branch-1.3` there is nothing to notice from that! it is absolutely speedy and usable!**

Wonder what has changes in FLTK 1.4 that makes it so slow.

**Comment by erco77** [comment]
@MatthiasWM Thanks for the description of operation -- I take it all those wrap calculations are to determine how many actual "scrollable" lines are in the buffer to figure out how to size the vertical scrollbar.

It sure sounds like caching these calculations in a per-line array would help to avoid redoing them. This way each line could have a cache of the vertical size of the line, based on the current wrap margin. So if a single line wrapped 3 times, it would have a height reflecting the sum height of those wraps. And the cache would only have to be updated if (a) the window is resized, (b) if a line is edited, only that line's cache would need updating. This way the overall height of the entire document is a quick walk of that cache.

It's probably more complicated than that, though..

@wcout I can only guess fixing a bug in the scrollbar calculations caused this (maybe?).
It's interesting that you're saying you loaded a 41,000 line file with wrap mode, and get bad behavior. I didn't actually know there were slowness problems outside of how Fl_Simple_Terminal abuses the editor as a terminal window (lots of fast and large append() operations). The editor probably needs a complete rewrite.

**Comment by wcout** [comment]
> @erco77 I can only guess fixing a bug in the scrollbar calculations caused this (maybe?).

I think you are right about this. It seems it has to do with changes for [STR 3412](https://www.fltk.org/str.php?L3412+P0+S0+C0+I0+E0+V%25+Q3412).

If I change the `oldTaWidth` initialization back to `text_area.w` as (before STR-3412) the full speed is there again in wrap mode.
Of course the bug (`display-bug.cxx` from STR-3312) is also there again :(

This was my change:

````diff
diff --git a/src/Fl_Text_Display.cxx b/src/Fl_Text_Display.cxx
index 00a4e6a7d..2ebc7dd7d 100644
--- a/src/Fl_Text_Display.cxx
+++ b/src/Fl_Text_Display.cxx
@@ -525,7 +525,7 @@ void Fl_Text_Display::recalc_display() {
   // Note: active since Oct 25, 2017: commit eb772d027d (svn r12526)
 
   // force _first_ calculation in loop (STR #3412)
-  int oldTAWidth = -1; // was: text_area.w (before STR #3412)
+//  int oldTAWidth = -1; // was: text_area.w (before STR #3412)
 
   if (mContinuousWrap && !mWrapMarginPix) {
 
@@ -539,6 +539,8 @@ void Fl_Text_Display::recalc_display() {
   }
   // End of optimization, see comment above.
 
+  int oldTAWidth = text_area.w; // (before STR #3412)
+
   for (int again = 1; again;) {
     again = 0;
     /* In continuous wrap mode, a change in width affects the total number of

````

**Comment by erco77** [comment]
@wcout Yes, I think Manolo came to the same conclusion you just did -- that same line -- in [issue 300](https://github.com/fltk/fltk/issues/300). His 3rd comment in that issue identifies a particular commit that made that change that caused the problem but fixed a bug. There was, AFAIK, no clear conclusion to how to fix both issues, at least not without a lot of retooling. My suggestion has been to cache per-line the computed line widths so that they don't have to be redone on a per-char basis, as apparently the additions for unicode made this non-trivial.

Meanwhile I'm retooling Fl_Simple_Terminal (rewriting it actually) to use its own Unicode text drawing code.. it's been keeping me busy for the last week, lol. Making good progress and think I have a clean design that can perhaps be used for rewriting Fl_Text_Editor.. will see.

**Comment by MatthiasWM** [comment]
I tried X11/Linux and scrolling a medium sized wrapping text is close to impossible. I believe that the main loop is slammed with drag events, keeping the widget busy recalculating the text sizes while never actually drawing them! If that is the case, it would explain a lot of the trouble s that we are having.

**Comment by Albrecht-S** [comment]
ISTR that the line width is calculated character by character. This is not only inexact but takes way too much time. If this is still true I would suggest to calculate width word by word, always including the leading space (' ') if any. This would probably reduce width calculation time massively (or maybe not that much, I don't know).

We could even calculate the entire line and use binary search or even some kind of interpolation to find the wrapping point.
1. Start by measuring the whole line. If it fits we're done, else continue.
2. Let's assume, the full line is measured at about 150% of the display width, then we measure 2/3 of the characters of the line to see if it fits. If not, interpolate again, or add or  remove one word to find the wrapping point. Then measure again.
3. Taking into account only word breaks would likely speed this up, particularly because we're only interested in word breaks for wrapping anyway.

This all would assume that measuring a word of 10 characters takes significantly less time than measuring all 10 characters one by one. I'm not sure if this assumption is true, but the algorithm of interpolation or binary search should speed this up anyway. The description above is not meant as an algorithm for direct use, it's just a model for improvement.

Even if we only measured whole words instead of single characters finding the line breaks should be faster.

**Comment by MatthiasWM** [comment]
So I just commited a change that should fix all issues with very minor drawbacks. 

Most of the time is wasted measuring text segments over and over. But much of the resut is merely used to calculate the scroll bar size, which is not precise anyway. So my patch instead estimates the number of lines using an educated guess. The result is magnitudes faster.

The drawback is an occasional scrolling of two lines when using the up or down arrow, but I think that's tollerable.

I have tested this on a Mac M1 with 450kB of text and it works just fine. At 1.5MB text, the slider feels slightly sticky. 2MB is still doable. I was able to edit an 11MB text file, but now loading the file is the new bottle neck at 6 seconds.

Please let me know if this change works well in your tests, or if you find bug or other issues.

**Comment by Albrecht-S** [comment]
I did a quick test with file size 4 MB and compared with FLTK 1.3. AFAICT the 1.4 version is still much slower than the 1.3 version. I can't tell how much the speed was improved in comparison to the state before your commit though. I'll try to test more, but I wanted to let you know that someone is testing... ;-)

**Comment by MatthiasWM** [comment]
The problem appears in 1.4 pre commit only when in wrapping mode, and with a 4MB file, wrapping should be *very* slow.

**Comment by Albrecht-S** [comment]
Yes, indeed, it is, that's why I tested it with wrapping mode set before doing anything else, in both 1.3 (Git) and 1.4 (Git). The point is that FLTK 1.3 loads the file much faster and scrolling is much smoother.

In other words: the current behavior in 1.4 (even with your latest commit) looks like a regression WRT 1.3, and that's what I'm worried about. I'm investigating further...

**Comment by Albrecht-S** [comment]
OK, more tests (all with wrap-mode ON) revealed:
1. commit 9bb9cb3f9 improves speed **more** than significantly
2. scrolling an 8 MiB file running under `perf` is almost unusable pre-9bb9cb3f9
3. scrolling the display is slower in 1.4 (after 9bb9cb3f9) than in 1.3, but somehow "acceptable"
4. resizing the window in 1.4 (after 9bb9cb3f9) is **much** faster than in old 1.3 (git)

My conclusion:
- commit 9bb9cb3f9 is a big improvement and seems to work well.

FWIW, if someone is interested: I'm posting an excerpt of two performance reports, i.e. with (a) current master and (b) pre-9bb9cb3f9. Note that I had to kill (`CTRL/C`) the operation when run with pre-9bb9cb3f9 because it didn't seem to finish (100% CPU), i.e. the second part of the report shows **_much more CPU activity for much less "action"_**. Test scenario: run `test/editor` with an 8MB file and scroll down to the end dragging the scrollbar, then release the mouse button. Note that the window doesn't get updated while dragging but at some time (seconds) after releasing the mouse button the scrollbar "jumps" to the end.

[perf-report.txt](https://github.com/user-attachments/files/16573795/perf-report.txt)

My interpretation:
- with current master the main CPU time (~87%) is consumed in two functions: `Fl_Text_Buffer::estimate_lines` and `Fl_Text_Buffer::count_lines`. These functions are "relatively quick" because they don't call too much graphics stuff.
- with pre-9bb9cb3f9 code there are a lot more functions involved that iterate through the buffer and measure strings which is **much** slower.

Finally: @MatthiasWM Great job, the result is a big improvement. Files smaller than 1-2 MB are pretty well usable on current hardware, files bigger than 4-8 MB have issues when dragging the scrollbar but I think we can live with this. Thanks a lot!

**Comment by Albrecht-S** [comment]
FWIW, some observations using `gedit` [1] and `scite` [2] in wrap mode under Gnome: both editors seem to update their scrollbar positions (i.e. the position of the slider) in the background
- when loading the file and
- when resizing the window.

The scrollbar's handle (slider) seems to keep its size constant when it's at some minimal size but you can see it move up and down while loading the file and while resizing the window. This keeps the interface responsive while updating the scrollbar position. Not a bad idea.

Note: this is just an observation, I don't propose to do this as well in FLTK 1.4 but it might be an option in 1.5 ...

--------------------------
[1] in `gedit` wrap mode is enabled by default
[2] use `scite -wrap=1 test.txt` to enable wrap mode

**Comment by Albrecht-S** [comment]
Visual Studio Code opening an 18 MiB file:
![image](https://github.com/user-attachments/assets/5f61d0af-4137-4e1e-9c6c-a08467186c9e)

I did not yet try to "Forcefully Enable Features" ;-)
I like "to ... avoid freezing or crashing". ;-)

**Comment by wcout** [comment]
> @Albrecht-S  Files smaller than 1-2 MB are pretty well usable on current hardware, files bigger than 4-8 MB have issues when dragging the scrollbar but I think we can live with this. Thanks a lot!

FWIW: The patch I tried for issue #300 does also improve this scrolling latency:

[textdisplay_trigger_recalc_display_v4.diff.txt](https://github.com/user-attachments/files/16575513/textdisplay_trigger_recalc_display_v4.diff.txt)

Scrolling of a 6MB file is seamless.

**Comment by MatthiasWM** [comment]
Thanks for the patch. That really helps when using the scroll bar. Happy to finally close this issue!

**Comment by Albrecht-S** [comment]
Yeah, thanks to @wcout for the patch. FTR: big files still experience "stuttering" if the scrollbar is dragged sloooowly (as expected because of multiple `draw()` calls) but it's really quick if the scrollbar is dragged quickly to a point (e.g. to the end) and then released (less `draw()` calls).
